### PR TITLE
Firstboot fixes and refactoring

### PR DESCRIPTION
**Pull request by SunilMohanAdapa**
- Fix major regression so that steps can submitted multiple times in
  case of errors.

- Don't serve the welcome page (and other pages) only once.  Show it
  until action is an taken.  This does not apply to the final step.

- Eliminate all coupling of one first boot step on another.

- Move first boot helper methods to __init__.py instead of middleware as
  it is more generic than middleware.

- Implement caching the first boot state to avoid an SQL query on every
  page load.  The down side is that if first boot state is modified in
  the backend DB outside Plinth, Plinth will need to be restarted to
  catch the modified value.

- Mark some methods as private.

- Refactor middleware code for slightly more simplicity.

- Don't show sidebar in pagekite first boot step.  Set width like other
  pages.

- pagekite: Fix first boot to use requests module

- Rename views/urls to be non-numeric

- Indentation & styling fixes

### CODE REVIEW

**Comment by SunilMohanAdapa** [comment]
@harry-7 , please see these changes which are in addition to your earlier pull request.

**Comment by jvalleroy** [comment]
@SunilMohanAdapa I'm testing this in vagrant (non-Danube edition), and noticed a couple issues:
1. After creating the 1st user, I'm still stuck at the create user page. I had to create a 2nd user to go on.
2. Then it took me to the module setup page for pagekite. I had to complete the pagekite module setup, and then I got to the setup complete screen.

File: plinth/modules/users/views.py [hunk 1]
```diff
@@ -168,22 +167,25 @@ def form_valid(self, form):
...
         """Make request available to the form (to insert messages)"""
-        kwargs = super(State1View, self).get_form_kwargs()
+        kwargs = super(FirstBootView, self).get_form_kwargs()
         kwargs['request'] = self.request
         return kwargs
+
+    def form_valid(self, form):
+        """Mark this first boot step as completed and save form."""
+        if User.objects.all():
+            first_boot.mark_step_done('users_firstboot')
```
Comment (jvalleroy) [inline plinth/modules/users/views.py]: It looks like form_valid is called before the user is created. We can move this line into FirstBootForm.save().

**Comment by jvalleroy** [comment]
Regarding the pagekite issue, I thought of 2 possible solutions:
- Require that any module declaring first_boot_steps must be an essential module.
- Or, check for each first_boot_step that its module has been setup, and skip it not setup yet.

**Comment by jvalleroy** [comment]
I've merged this, with additional changes:
- users: Fix marking first boot step as done
- firstboot: Only add steps from modules if they have completed setup

**Pull request closed without merging**